ToolBox/PsychSignalDataConverter/PsychSignalDataDownloader.cs

```diff
@@ -0,0 +1,149 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.IO;
+using System.Net;
+using System.Threading;
+using NodaTime;
+using QuantConnect.Logging;
+
+namespace QuantConnect.ToolBox.PsychSignalDataConverter
+{
+    public class PsychSignalDataDownloader
+    {
+        private readonly string _apiKey;
+        private readonly string _dataSource;
+        
+        /// <summary>
+        /// Base URL for the psychsignal API
+        /// </summary>
+        private readonly string _baseUrl = "https://api.psychsignal.com/v2";
+
+        /// <summary>
+        /// Destination we will write raw data to
+        /// </summary>
+        private readonly string _rawDataDestination;
+        
+        /// <summary>
+        /// Maximum amount of retries per data hour 
+        /// </summary>
+        public int MaxRetries = 5;
+        
+        /// <summary>
+        /// Downlods data from psychsignal
+        /// </summary>
+        /// <param name="rawDataDestination">Directory we write raw data to</param>
+        /// <param name="apiKey">PsychSignal API key</param>
+        /// <param name="dataSource">Data source (e.g. stocktwits,twitter_withretweets)</param>
+        public PsychSignalDataDownloader(string rawDataDestination, string apiKey, string dataSource)
+        {
+            _rawDataDestination = rawDataDestination;
+
+            _dataSource = dataSource;
+            _apiKey = apiKey;
+        }
+        
+        /// <summary>
+        /// Download the data from the given starting date to the ending date.
+        /// Note that if the ending date is in the same hour as the current time,
+        /// we will lower the <paramref name="endDateUtc" /> by one hour in order
+        /// to make sure that we only download complete, non-changing dataa
+        /// </summary>
+        /// <param name="startDateUtc">Starting date</param>
+        /// <param name="endDateUtc">Ending date</param>
+        public void Download(DateTime startDateUtc, DateTime endDateUtc)
+        {
+            if (startDateUtc < endDateUtc.AddDays(-15))
+            {
+                throw new ArgumentException("The starting date can only be at most 15 days from now");
```

Comment:
> "The starting date can only be at most 15 days **from now"**

I think it should say from end date?

Reply:
PsychSignal only allows you to reference data from 15 hours from the current time

Reply:
Right, my comment wasn't that clear but I was pointing at the fact that the code seems to asume `endDateUtc` is ~now? Maybe it should directly check start date against `UtcNow.AddDays(-15)`